File: backend/app/auth/authority_validation.py
"""
ThermoShield Authority Access Request Validation
Canonical compatibility validation between Organization, Department,
Designation, Functional Role, and Operational Jurisdiction.
"""
from typing import Tuple, List, Dict, Any
import logging

logger = logging.getLogger(__name__)

# Configured Authority Organizations and their allowed parameters
CONFIGURED_ORGANIZATIONS = {
    "MCGM": {
        "id": "MCGM",
        "name": "Municipal Corporation of Greater Mumbai (MCGM / BMC)",
        "allowed_levels": ["MUNICIPAL_CORPORATION", "ADMINISTRATIVE_WARD"],
        "primary_jurisdiction": "IN-MH-MCGM",
        "allowed_roles": ["municipal_hap_officer", "ward_officer", "responder", "official"],
        "departments": [
            "Disaster Management Cell",
            "Heat Action / Climate Cell",
            "Public Health",
            "Urban Health",
            "Administration",
        ],
    },
    "MH_SDMA": {
        "id": "MH_SDMA",
        "name": "Maharashtra State Disaster Management Authority (SDMA)",
        "allowed_levels": ["STATE_UT"],
        "primary_jurisdiction": "IN-MH",
        "allowed_roles": ["state_coordinator", "national_analyst", "responder", "official"],
        "departments": [
            "Disaster Management Cell",
            "State Emergency Operations Centre",
            "Heat Action / Climate Cell",
            "Emergency Response",
            "Administration",
        ],
    },
    "MH_PHD": {
        "id": "MH_PHD",
        "name": "Maharashtra Public Health Department",
        "allowed_levels": ["STATE_UT", "DISTRICT"],
        "primary_jurisdiction": "IN-MH",
        "allowed_roles": ["state_coordinator", "district_authority", "official"],
        "departments": [
            "Public Health",
            "Urban Health",
            "Administration",
            "Emergency Response",
        ],
    },
    "PUNE_DDMA": {
        "id": "PUNE_DDMA",
        "name": "District Disaster Management Authority — Pune",
        "allowed_levels": ["DISTRICT"],
        "primary_jurisdiction": "IN-MH-DIST-PUN",
        "allowed_roles": ["district_authority", "responder", "official"],
        "departments": [
            "Disaster Management Cell",
            "Emergency Response",
            "Public Health",
            "Administration",
        ],
    },
    "NAGPUR_DDMA": {
        "id": "NAGPUR_DDMA",
        "name": "District Disaster Management Authority — Nagpur",
        "allowed_levels": ["DISTRICT"],
        "primary_jurisdiction": "IN-MH-DIST-NAGPUR",
        "allowed_roles": ["district_authority", "responder", "official"],
        "departments": [
            "Disaster Management Cell",
            "Emergency Response",
            "Public Health",
            "Administration",
        ],
    },
    "NDMA": {
        "id": "NDMA",
        "name": "National Disaster Management Authority (NDMA)",
        "allowed_levels": ["COUNTRY"],
        "primary_jurisdiction": "IN",
        "allowed_roles": ["national_analyst", "system_admin", "responder", "official"],
        "departments": [
            "Disaster Management Cell",
            "Emergency Response",
            "Administration",
        ],
    },
    "IMD": {
        "id": "IMD",
        "name": "India Meteorological Department (IMD)",
        "allowed_levels": ["COUNTRY"],
        "primary_jurisdiction": "IN",
        "allowed_roles": ["national_analyst", "analyst"],
        "departments": [
            "Meteorological Analysis",
            "Heat Action / Climate Cell",
            "Administration",
        ],
    },
    "NCDC": {
        "id": "NCDC",
        "name": "National Centre for Disease Control (NCDC)",
        "allowed_levels": ["COUNTRY"],
        "primary_jurisdiction": "IN",
        "allowed_roles": ["national_analyst", "district_authority", "official"],
        "departments": [
            "Public Health",
            "Urban Health",
            "Administration",
        ],
    },
}

ROLE_KEY_NORMALIZATION = {
    "heat action plan officer": "municipal_hap_officer",
    "municipal_hap_officer": "municipal_hap_officer",
    "heat-risk analyst": "national_analyst",
    "national_analyst": "national_analyst",
    "public health monitoring": "district_authority",
    "district_authority": "district_authority",
    "emergency / field response": "responder",
    "responder": "responder",
    "ward operations": "ward_officer",
    "ward_officer": "ward_officer",
    "state coordination": "state_coordinator",
    "state_coordinator": "state_coordinator",
    "national analysis": "national_analyst",
    "system_admin": "system_admin",
    "official": "official",
}


def match_organization(org_str: str | None) -> Dict[str, Any] | None:
    if not org_str:
        return None
    org_upper = org_str.strip().upper()

    # Exact key match
    if org_upper in CONFIGURED_ORGANIZATIONS:
        return CONFIGURED_ORGANIZATIONS[org_upper]

    # Substring / Keyword matching
    if "MCGM" in org_upper or "GREATER MUMBAI" in org_upper or "BMC" in org_upper:
        return CONFIGURED_ORGANIZATIONS["MCGM"]
    if "SDMA" in org_upper or ("MAHARASHTRA" in org_upper and "DISASTER" in org_upper):
        return CONFIGURED_ORGANIZATIONS["MH_SDMA"]
    if "MAHARASHTRA" in org_upper and ("HEALTH" in org_upper or "PHD" in org_upper):
        return CONFIGURED_ORGANIZATIONS["MH_PHD"]
    if "PUNE" in org_upper and "DDMA" in org_upper:
        return CONFIGURED_ORGANIZATIONS["PUNE_DDMA"]
    if "NAGPUR" in org_upper and "DDMA" in org_upper:
        return CONFIGURED_ORGANIZATIONS["NAGPUR_DDMA"]
    if "NDMA" in org_upper or "NATIONAL DISASTER" in org_upper:
        return CONFIGURED_ORGANIZATIONS["NDMA"]
    if "IMD" in org_upper or "METEOROLOGICAL" in org_upper:
        return CONFIGURED_ORGANIZATIONS["IMD"]
    if "NCDC" in org_upper or "DISEASE CONTROL" in org_upper:
        return CONFIGURED_ORGANIZATIONS["NCDC"]

    return None


def validate_authority_access_request(
    organization: str | None,
    department: str | None,
    designation: str | None,
    requested_role: str | None,
    jurisdiction_id: str | None,
    jurisdiction_type: str | None = None,
) -> Tuple[bool, List[str], Dict[str, Any]]:
    """
    Validates mutually consistent Organization, Department, Designation,
    Functional Role, and Jurisdiction.

    Returns:
        (is_valid, errors_list, normalized_data_dict)
    """
    errors: List[str] = []
    normalized: Dict[str, Any] = {
        "organization": organization,
        "department": department,
        "designation": designation,
        "normalized_role": "user",
        "normalized_jurisdiction_id": jurisdiction_id or "IN",
        "normalized_jurisdiction_type": jurisdiction_type or "COUNTRY",
        "normalized_jurisdiction_name": "India (National)",
    }

    if not organization or not organization.strip():
        errors.append("Organization is required for authority access requests.")
        return False, errors, normalized

    org_config = match_organization(organization)
    if not org_config:
        errors.append(f"Organization '{organization}' is not recognized in the configured authority catalog.")
        return False, errors, normalized

    normalized["organization"] = org_config["name"]

    # Normalize Role
    role_input = (requested_role or "").strip().lower()
    canonical_role = ROLE_KEY_NORMALIZATION.get(role_input, role_input)
    normalized["normalized_role"] = canonical_role

    # Verify role is allowed for organization
    if canonical_role not in org_config["allowed_roles"]:
        allowed_names = ", ".join(org_config["allowed_roles"])
        errors.append(
            f"Role '{requested_role}' is not compatible with '{org_config['name']}'. "
            f"Allowed roles: {allowed_names}."
        )

    # Resolve and Validate Jurisdiction
    req_juris = (jurisdiction_id or "").strip()
    if not req_juris:
        errors.append("Operational jurisdiction is required for authority access requests.")
        return False, errors, normalized

    normalized["normalized_jurisdiction_id"] = req_juris

    # Specific Organization Jurisdiction Rules
    org_id = org_config["id"]
    if org_id == "MCGM":
        # Must be IN-MH-MCGM or a BMC ward
        if req_juris != "IN-MH-MCGM" and not req_juris.startswith("IN-MH-MCGM-"):
            errors.append(
                f"MCGM operates strictly within Greater Mumbai (IN-MH-MCGM) or its administrative wards. "
                f"Requested jurisdiction '{req_juris}' is invalid."
            )
        else:
            normalized["normalized_jurisdiction_type"] = (
                "ADMINISTRATIVE_WARD" if req_juris.startswith("IN-MH-MCGM-") else "MUNICIPAL_CORPORATION"
            )
            normalized["normalized_jurisdiction_name"] = (
                f"BMC Ward ({req_juris.replace('IN-MH-MCGM-', '')})"
                if req_juris.startswith("IN-MH-MCGM-")
                else "Greater Mumbai (MCGM)"
            )

    elif org_id in ("MH_SDMA", "MH_PHD"):
        # Must be IN-MH or a Maharashtra district
        if req_juris != "IN-MH" and not req_juris.startswith("IN-MH-DIST-"):
            errors.append(
                f"'{org_config['name']}' operates within Maharashtra (IN-MH) or its subordinate districts. "
                f"Requested jurisdiction '{req_juris}' is invalid."
            )
        else:
            normalized["normalized_jurisdiction_type"] = (
                "DISTRICT" if req_juris.startswith("IN-MH-DIST-") else "STATE_UT"
            )
            normalized["normalized_jurisdiction_name"] = (
                f"District ({req_juris.replace('IN-MH-DIST-', '')})"
                if req_juris.startswith("IN-MH-DIST-")
                else "Maharashtra"
            )

    elif org_id == "PUNE_DDMA":
        if req_juris != "IN-MH-DIST-PUN" and "PUN" not in req_juris.upper():
            errors.append(
                f"Pune DDMA operates strictly within Pune District (IN-MH-DIST-PUN). "
                f"Requested jurisdiction '{req_juris}' is invalid."
            )
        else:
            normalized["normalized_jurisdiction_id"] = "IN-MH-DIST-PUN"
            normalized["normalized_jurisdiction_type"] = "DISTRICT"
            normalized["normalized_jurisdiction_name"] = "Pune District"

    elif org_id == "NAGPUR_DDMA":
        if req_juris != "IN-MH-DIST-NAGPUR" and "NAGPUR" not in req_juris.upper():
            errors.append(
                f"Nagpur DDMA operates strictly within Nagpur District (IN-MH-DIST-NAGPUR). "
                f"Requested jurisdiction '{req_juris}' is invalid."
            )
        else:
            normalized["normalized_jurisdiction_id"] = "IN-MH-DIST-NAGPUR"
            normalized["normalized_jurisdiction_type"] = "DISTRICT"
            normalized["normalized_jurisdiction_name"] = "Nagpur District"

    elif org_id in ("NDMA", "IMD", "NCDC"):
        if req_juris != "IN":
            errors.append(
                f"'{org_config['name']}' operates at the National level (IN). "
                f"Requested jurisdiction '{req_juris}' is invalid."
            )
        else:
            normalized["normalized_jurisdiction_id"] = "IN"
            normalized["normalized_jurisdiction_type"] = "COUNTRY"
            normalized["normalized_jurisdiction_name"] = "India (National)"

    # Role <-> Jurisdiction semantic consistency
    if canonical_role == "municipal_hap_officer" and normalized["normalized_jurisdiction_type"] != "MUNICIPAL_CORPORATION":
        errors.append("Municipal HAP Nodal Officer role must be assigned to a Municipal Corporation jurisdiction (e.g. IN-MH-MCGM).")

    if canonical_role == "ward_officer" and normalized["normalized_jurisdiction_type"] != "ADMINISTRATIVE_WARD":
        errors.append("Ward Officer role must be assigned to an Administrative Ward jurisdiction.")

    if canonical_role == "state_coordinator" and normalized["normalized_jurisdiction_type"] != "STATE_UT":
        errors.append("State Coordinator role must be assigned to a State / UT jurisdiction (e.g. IN-MH).")

    if canonical_role == "national_analyst" and normalized["normalized_jurisdiction_type"] not in ("COUNTRY", "STATE_UT"):
        errors.append("National / Regional Analyst role cannot be assigned below State level.")

    is_valid = len(errors) == 0
    return is_valid, errors, normalized

File: frontend/src/utils/authorityValidation.ts
/**
 * Authority Access Request Compatibility Validator
 * Ensures mutual consistency across Organization, Department, Designation,
 * Functional Role, Administrative Level, and Operational Jurisdiction.
 */

export interface AuthorityValidationResult {
  valid: boolean;
  errors: string[];
  normalizedRole: string;
  normalizedJurisdictionId: string;
  normalizedJurisdictionName: string;
  normalizedJurisdictionType: string;
  defaultPermissions: string[];
}

export interface AuthorityRequestParams {
  organization?: string;
  department?: string;
  designation?: string;
  requestedRole?: string;
  jurisdictionType?: string;
  jurisdictionId?: string;
}

import { INDIAN_MUNICIPAL_CORPORATIONS } from '../data/indianMunicipalCorporations';

interface OrgRule {
  id: string;
  name: string;
  allowedLevels: string[];
  primaryJurisdiction: string;
  allowedRoles: string[];
  isMunicipal?: boolean;
}

const CONFIGURED_ORGS: Record<string, OrgRule> = {
  MCGM: {
    id: 'MCGM',
    name: 'Municipal Corporation of Greater Mumbai (MCGM / BMC)',
    allowedLevels: ['MUNICIPAL_CORPORATION', 'ADMINISTRATIVE_WARD'],
    primaryJurisdiction: 'IN-MH-MCGM',
    allowedRoles: ['municipal_hap_officer', 'ward_officer', 'responder', 'official'],
    isMunicipal: true,
  },
  MH_SDMA: {
    id: 'MH_SDMA',
    name: 'Maharashtra State Disaster Management Authority (SDMA)',
    allowedLevels: ['STATE_UT'],
    primaryJurisdiction: 'IN-MH',
    allowedRoles: ['state_coordinator', 'national_analyst', 'responder', 'official'],
  },
  MH_PHD: {
    id: 'MH_PHD',
    name: 'Maharashtra Public Health Department',
    allowedLevels: ['STATE_UT', 'DISTRICT'],
    primaryJurisdiction: 'IN-MH',
    allowedRoles: ['state_coordinator', 'district_authority', 'official'],
  },
  PUNE_DDMA: {
    id: 'PUNE_DDMA',
    name: 'District Disaster Management Authority — Pune',
    allowedLevels: ['DISTRICT'],
    primaryJurisdiction: 'IN-MH-DIST-PUN',
    allowedRoles: ['district_authority', 'responder', 'official'],
  },
  NAGPUR_DDMA: {
    id: 'NAGPUR_DDMA',
    name: 'District Disaster Management Authority — Nagpur',
    allowedLevels: ['DISTRICT'],
    primaryJurisdiction: 'IN-MH-DIST-NAGPUR',
    allowedRoles: ['district_authority', 'responder', 'official'],
  },
  NDMA: {
    id: 'NDMA',
    name: 'National Disaster Management Authority (NDMA)',
    allowedLevels: ['COUNTRY'],
    primaryJurisdiction: 'IN',
    allowedRoles: ['national_analyst', 'system_admin', 'responder', 'official'],
  },
  IMD: {
    id: 'IMD',
    name: 'India Meteorological Department (IMD)',
    allowedLevels: ['COUNTRY'],
    primaryJurisdiction: 'IN',
    allowedRoles: ['national_analyst', 'analyst'],
  },
  NCDC: {
    id: 'NCDC',
    name: 'National Centre for Disease Control (NCDC)',
    allowedLevels: ['COUNTRY'],
    primaryJurisdiction: 'IN',
    allowedRoles: ['national_analyst', 'district_authority', 'official'],
  },
};

const ROLE_CANONICAL_MAP: Record<string, string> = {
  'heat action plan officer': 'municipal_hap_officer',
  'municipal_hap_officer': 'municipal_hap_officer',
  'heat-risk analyst': 'national_analyst',
  'national_analyst': 'national_analyst',
  'public health monitoring': 'district_authority',
  'district_authority': 'district_authority',
  'emergency / field response': 'responder',
  'responder': 'responder',
  'ward operations': 'ward_officer',
  'ward_officer': 'ward_officer',
  'state coordination': 'state_coordinator',
  'state_coordinator': 'state_coordinator',
  'national analysis': 'national_analyst',
  'system_admin': 'system_admin',
  'official': 'official',
};

const DEFAULT_ROLE_PERMS: Record<string, string[]> = {
  national_analyst: ['VIEW_JURISDICTION', 'VIEW_NATIONAL_CONTEXT', 'VIEW_SUBORDINATE_REGIONS', 'ANALYZE_RISK', 'RECOMMEND_HAP_ACTION', 'EXPORT_REPORT'],
  state_coordinator: ['VIEW_JURISDICTION', 'VIEW_PARENT_CONTEXT', 'VIEW_NATIONAL_CONTEXT', 'VIEW_SUBORDINATE_REGIONS', 'ANALYZE_RISK', 'RECOMMEND_HAP_ACTION', 'EXPORT_REPORT'],
  district_authority: ['VIEW_JURISDICTION', 'VIEW_PARENT_CONTEXT', 'VIEW_SUBORDINATE_REGIONS', 'ANALYZE_RISK', 'RECOMMEND_HAP_ACTION', 'ACTIVATE_HAP', 'APPROVE_HAP_ACTION', 'DISPATCH_RESPONDER', 'EXPORT_REPORT'],
  municipal_hap_officer: ['VIEW_JURISDICTION', 'VIEW_PARENT_CONTEXT', 'VIEW_NATIONAL_CONTEXT', 'VIEW_SUBORDINATE_REGIONS', 'ANALYZE_RISK', 'RECOMMEND_HAP_ACTION', 'APPROVE_HAP_ACTION', 'ACTIVATE_HAP', 'CLOSE_HAP_ACTION', 'SEND_PUBLIC_ADVISORY', 'DISPATCH_RESPONDER', 'EXPORT_REPORT'],
  ward_officer: ['VIEW_JURISDICTION', 'VIEW_PARENT_CONTEXT', 'ACKNOWLEDGE_TASK', 'DISPATCH_RESPONDER', 'RECOMMEND_HAP_ACTION'],
  responder: ['VIEW_JURISDICTION', 'ACKNOWLEDGE_TASK'],
  system_admin: ['MANAGE_JURISDICTION_USERS', 'VIEW_JURISDICTION'],
};

function matchOrg(orgStr?: string): OrgRule | null {
  if (!orgStr) return null;
  const upper = orgStr.trim().toUpperCase();
  if (CONFIGURED_ORGS[upper]) return CONFIGURED_ORGS[upper];

  // Match against 50+ Indian Municipal Corporations Catalog
  const mcMatch = INDIAN_MUNICIPAL_CORPORATIONS.find(
    (c) =>
      c.shortCode.toUpperCase() === upper ||
      c.id.toUpperCase() === upper ||
      c.name.toUpperCase() === upper ||
      upper.includes(c.name.toUpperCase()) ||
      (upper.includes(c.city.toUpperCase()) && (upper.includes('MUNICIPAL') || upper.includes('NAGAR NIGAM') || upper.includes('CORPORATION') || upper.includes('MAHANAGARA')))
  );

  if (mcMatch) {
    return {
      id: mcMatch.shortCode,
      name: mcMatch.name,
      allowedLevels: ['MUNICIPAL_CORPORATION', 'ADMINISTRATIVE_WARD'],
      primaryJurisdiction: mcMatch.id,
      allowedRoles: ['municipal_hap_officer', 'ward_officer', 'responder', 'official'],
      isMunicipal: true,
    };
  }

  if (upper.includes('MCGM') || upper.includes('GREATER MUMBAI') || upper.includes('BMC')) return CONFIGURED_ORGS.MCGM;
  if (upper.includes('SDMA') || (upper.includes('MAHARASHTRA') && upper.includes('DISASTER'))) return CONFIGURED_ORGS.MH_SDMA;
  if (upper.includes('MAHARASHTRA') && (upper.includes('HEALTH') || upper.includes('PHD'))) return CONFIGURED_ORGS.MH_PHD;
  if (upper.includes('PUNE') && upper.includes('DDMA')) return CONFIGURED_ORGS.PUNE_DDMA;
  if (upper.includes('NAGPUR') && upper.includes('DDMA')) return CONFIGURED_ORGS.NAGPUR_DDMA;
  if (upper.includes('NDMA') || upper.includes('NATIONAL DISASTER')) return CONFIGURED_ORGS.NDMA;
  if (upper.includes('IMD') || upper.includes('METEOROLOGICAL')) return CONFIGURED_ORGS.IMD;
  if (upper.includes('NCDC') || upper.includes('DISEASE CONTROL')) return CONFIGURED_ORGS.NCDC;

  return null;
}

export function validateAuthorityAccessRequest(params: AuthorityRequestParams): AuthorityValidationResult {
  const errors: string[] = [];
  const { organization, department, designation, requestedRole, jurisdictionId } = params;

  let normalizedRole = 'user';
  let normalizedJurisdictionId = jurisdictionId || 'IN';
  let normalizedJurisdictionName = 'India (National)';
  let normalizedJurisdictionType = 'COUNTRY';

  if (!organization || !organization.trim()) {
    errors.push('Government organization is required.');
    return {
      valid: false,
      errors,
      normalizedRole,
      normalizedJurisdictionId,
      normalizedJurisdictionName,
      normalizedJurisdictionType,
      defaultPermissions: [],
    };
  }

  const orgConfig = matchOrg(organization);
  if (!orgConfig) {
    errors.push(`Organization '${organization}' is not recognized in the configured catalog.`);
    return {
      valid: false,
      errors,
      normalizedRole,
      normalizedJurisdictionId,
      normalizedJurisdictionName,
      normalizedJurisdictionType,
      defaultPermissions: [],
    };
  }

  // Role Normalization
  const roleInput = (requestedRole || '').trim().toLowerCase();
  normalizedRole = ROLE_CANONICAL_MAP[roleInput] || roleInput;

  if (!orgConfig.allowedRoles.includes(normalizedRole)) {
    const allowed = orgConfig.allowedRoles.map((r) => r.replace(/_/g, ' ')).join(', ');
    errors.push(`Role '${requestedRole}' is incompatible with ${orgConfig.name}. Compatible roles: ${allowed}.`);
  }

  // Jurisdiction checks
  const reqJuris = (jurisdictionId || '').trim();
  if (!reqJuris) {
    errors.push('Operational jurisdiction is required.');
    return {
      valid: false,
      errors,
      normalizedRole,
      normalizedJurisdictionId,
      normalizedJurisdictionName,
      normalizedJurisdictionType,
      defaultPermissions: [],
    };
  }

  normalizedJurisdictionId = reqJuris;

  if (orgConfig.isMunicipal || orgConfig.id === 'MCGM') {
    const primaryId = orgConfig.primaryJurisdiction;
    const isPrimary = reqJuris.toUpperCase() === primaryId.toUpperCase();
    const isWard = reqJuris.toUpperCase().startsWith(`${primaryId.toUpperCase()}-`) || reqJuris.toLowerCase().startsWith('ward_');
    const isMatched = isPrimary || isWard || reqJuris.toUpperCase().includes(orgConfig.id.toUpperCase());

    if (!isMatched) {
      errors.push(`${orgConfig.name} operates strictly within its designated municipal corporation (${primaryId}) or subordinate wards. Requested jurisdiction '${reqJuris}' is invalid.`);
    } else {
      normalizedJurisdictionId = isWard ? reqJuris : primaryId;
      normalizedJurisdictionType = isWard ? 'ADMINISTRATIVE_WARD' : 'MUNICIPAL_CORPORATION';
      normalizedJurisdictionName = isWard
        ? `${orgConfig.name} Ward (${reqJuris.replace(`${primaryId}-`, '')})`
        : orgConfig.name;
    }
  } else if (orgConfig.id === 'MH_SDMA' || orgConfig.id === 'MH_PHD') {
    if (reqJuris !== 'IN-MH' && !reqJuris.startsWith('IN-MH-DIST-')) {
      errors.push(`'${orgConfig.name}' operates strictly within Maharashtra (IN-MH) or its subordinate districts. Requested jurisdiction '${reqJuris}' is invalid.`);
    } else {
      normalizedJurisdictionType = reqJuris.startsWith('IN-MH-DIST-') ? 'DISTRICT' : 'STATE_UT';
      normalizedJurisdictionName = reqJuris.startsWith('IN-MH-DIST-')
        ? `District (${reqJuris.replace('IN-MH-DIST-', '')})`
        : 'Maharashtra';
    }
  } else if (orgConfig.id === 'PUNE_DDMA') {
    if (reqJuris !== 'IN-MH-DIST-PUN' && !reqJuris.toUpperCase().includes('PUN')) {
      errors.push(`Pune DDMA operates strictly within Pune District (IN-MH-DIST-PUN). Requested jurisdiction '${reqJuris}' is invalid.`);
    } else {
      normalizedJurisdictionId = 'IN-MH-DIST-PUN';
      normalizedJurisdictionType = 'DISTRICT';
      normalizedJurisdictionName = 'Pune District';
    }
  } else if (orgConfig.id === 'NAGPUR_DDMA') {
    if (reqJuris !== 'IN-MH-DIST-NAGPUR' && !reqJuris.toUpperCase().includes('NAGPUR')) {
      errors.push(`Nagpur DDMA operates strictly within Nagpur District (IN-MH-DIST-NAGPUR). Requested jurisdiction '${reqJuris}' is invalid.`);
    } else {
      normalizedJurisdictionId = 'IN-MH-DIST-NAGPUR';
      normalizedJurisdictionType = 'DISTRICT';
      normalizedJurisdictionName = 'Nagpur District';
    }
  } else if (['NDMA', 'IMD', 'NCDC'].includes(orgConfig.id)) {
    if (reqJuris !== 'IN') {
      errors.push(`'${orgConfig.name}' operates at National level (IN). Requested jurisdiction '${reqJuris}' is invalid.`);
    } else {
      normalizedJurisdictionId = 'IN';
      normalizedJurisdictionType = 'COUNTRY';
      normalizedJurisdictionName = 'India (National)';
    }
  }

  // Cross-field role semantic alignment
  if (normalizedRole === 'municipal_hap_officer' && normalizedJurisdictionType !== 'MUNICIPAL_CORPORATION') {
    errors.push('Municipal HAP Nodal Officer role must be paired with a Municipal Corporation jurisdiction (e.g. Greater Mumbai).');
  }
  if (normalizedRole === 'ward_officer' && normalizedJurisdictionType !== 'ADMINISTRATIVE_WARD') {
    errors.push('Ward Officer role must be paired with an Administrative Ward jurisdiction.');
  }
  if (normalizedRole === 'state_coordinator' && normalizedJurisdictionType !== 'STATE_UT') {
    errors.push('State Coordinator role must be paired with a State / UT jurisdiction (e.g. Maharashtra).');
  }
  if (normalizedRole === 'national_analyst' && !['COUNTRY', 'STATE_UT'].includes(normalizedJurisdictionType)) {
    errors.push('National / Regional Analyst cannot be assigned below State level.');
  }

  const defaultPermissions = DEFAULT_ROLE_PERMS[normalizedRole] || ['VIEW_JURISDICTION'];

  return {
    valid: errors.length === 0,
    errors,
    normalizedRole,
    normalizedJurisdictionId,
    normalizedJurisdictionName,
    normalizedJurisdictionType,
    defaultPermissions,
  };
}
